stackhpc_monasca_agent_plugins/checks/nvidia.py

```diff
@@ -0,0 +1,225 @@
+        return {
+            'utilisation_gpu_percent': util.gpu,
+            'utilisation_memory_percent': util.memory
+        }
+
+    @staticmethod
+    @handle_not_supported
+    def _get_device_temperature(gpu):
+        return {'temperature_deg_c':
+                pynvml.nvmlDeviceGetTemperature(
+                    gpu, pynvml.NVML_TEMPERATURE_GPU)}
+
+    @staticmethod
+    @handle_not_supported
+    def _get_device_shutdown_temp(gpu):
+        return {'temperature_shutdown_deg_c':
+                pynvml.nvmlDeviceGetTemperatureThreshold(
+                    gpu, pynvml.NVML_TEMPERATURE_THRESHOLD_SHUTDOWN)}
+
+    @staticmethod
+    @handle_not_supported
+    def _get_device_slowdown_temp(gpu):
+        return {'temperature_slowdown_deg_c':
+                pynvml.nvmlDeviceGetTemperatureThreshold(
+                    gpu, pynvml.NVML_TEMPERATURE_THRESHOLD_SLOWDOWN)}
+
+    @staticmethod
+    @handle_not_supported
+    def _get_power_usage_watts(gpu):
+        return {'power_watts': (pynvml.nvmlDeviceGetPowerUsage(gpu) / 1000.0)}
+
+    @staticmethod
+    @handle_not_supported
+    def _get_power_limit_watts(gpu):
+        return {'power_limit_watts': (
+            pynvml.nvmlDeviceGetPowerManagementLimit(gpu) / 1000.0)}
+
+    @staticmethod
+    @handle_not_supported
+    def _get_clock_info(gpu):
+        return {
+            'clock_freq_gpu_mhz':
+                pynvml.nvmlDeviceGetClockInfo(gpu, pynvml.NVML_CLOCK_GRAPHICS),
+            'clock_freq_sm_mhz':
+                pynvml.nvmlDeviceGetClockInfo(gpu, pynvml.NVML_CLOCK_SM),
+            'clock_freq_memory_mhz':
+                pynvml.nvmlDeviceGetClockInfo(gpu, pynvml.NVML_CLOCK_MEM),
+            'clock_freq_video_mhz':
+                pynvml.nvmlDeviceGetClockInfo(gpu, pynvml.NVML_CLOCK_VIDEO)
+        }
+
+    @staticmethod
+    @handle_not_supported
+    def _get_clock_max_info(gpu):
+        return {
+            'clock_max_freq_gpu_mhz':
+                pynvml.nvmlDeviceGetMaxClockInfo(
+                    gpu, pynvml.NVML_CLOCK_GRAPHICS),
+            'clock_max_freq_sm_mhz':
+                pynvml.nvmlDeviceGetMaxClockInfo(gpu, pynvml.NVML_CLOCK_SM),
+            'clock_max_freq_memory_mhz':
+                pynvml.nvmlDeviceGetMaxClockInfo(gpu, pynvml.NVML_CLOCK_MEM),
+            'clock_max_freq_video_mhz':
+                pynvml.nvmlDeviceGetMaxClockInfo(gpu, pynvml.NVML_CLOCK_VIDEO)
+        }
+
+    @staticmethod
+    def _get_gpu_info():
+        pynvml.nvmlInit()
+        deviceCount = pynvml.nvmlDeviceGetCount()
+        all_info = []
+        for i in range(0, deviceCount):
+            gpu = pynvml.nvmlDeviceGetHandleByIndex(i)
+
+            dimensions = {}
+            dimensions.update(Nvidia._get_driver_version())
+            dimensions.update(Nvidia._get_device_uuid(gpu))
+            dimensions.update(Nvidia._get_info_rom_image_version(gpu))
+            dimensions.update(Nvidia._get_device_power_state(gpu))
```

Comment:
Will multiple time series be created if the power state is changed? Would that be better as metadata, or as a measurement?

Reply:
The dimensions are saved as tags in InfluxDB which are attached to the series for the metric in InfluxDB. The reason I included it as a dimension is because you might want to compare power consumption or temperature as a function of power state.

Reply:
What I mean is, isn't a single time series defined by a name and a set of dimensions and their values? So if we change the value of one of the dimensions over time, it effectively becomes a new time series?

Reply:
I know what you mean, but I think so long as the tag doesn't have high cardinality this is a normal way to handle things in InfluxDB. See the location tag in this example: https://docs.influxdata.com/influxdb/v1.6/concepts/key_concepts/#tag-set

The power state has 15 possible states. It's true that it could be a measurement. If it is a tag it's easy to plot average power consumption or temperature as a function of power state. If it's a measurement it's easy to see how it varies with time. Would you rather see it as a measurement?

Reply:
Perhaps my mental model was wrong - a series seems to be a much more fluid concept than I'd assumed. In which case I think this is fine.